src/Mod/Spreadsheet/Gui/SheetTableView.cpp

```diff
@@ -73,34 +74,91 @@ bool SheetViewHeader::viewportEvent(QEvent *e) {
-    connect(removeColumns, SIGNAL(triggered()), this, SLOT(removeColumns()));
-
-    QAction * cellProperties = new QAction(tr("Properties..."), this);
+    horizontalHeader()->setContextMenuPolicy(Qt::CustomContextMenu);
+    verticalHeader()->setContextMenuPolicy(Qt::CustomContextMenu);
+
+    connect(verticalHeader(), &QWidget::customContextMenuRequested,
+       [this](const QPoint &point){
+            QMenu menu(this);
+            const auto selection = selectionModel()->selectedRows();
+            const auto & [min, max] = selectedMinMaxRow(selection);
+            if (bool isContiguous = max - min == selection.size() - 1) {
+                /*: This is shown in the context menu for the vertical header in a spreadsheet.
+                    The number refers to how many lines are selected and will be inserted. */
+                auto insertBefore = menu.addAction(tr("Insert %n row(s) above", "", selection.size()));
+                connect(insertBefore, SIGNAL(triggered()), this, SLOT(insertRows()));
+    
+                if (max < model()->rowCount() - 1) {
+                    auto insertAfter = menu.addAction(tr("Insert %n row(s) below", "", selection.size()));
+                    connect(insertAfter, SIGNAL(triggered()), this, SLOT(insertRowsAfter()));
+                }
+            } else {
+                auto insert = menu.addAction(tr("Insert %n non-contiguous row(s)", "", selection.size()));
+                connect(insert, SIGNAL(triggered()), this, SLOT(insertRows()));
+            }
+            auto remove = menu.addAction(tr("Remove row(s)", "", selection.size()));
+            connect(remove, SIGNAL(triggered()), this, SLOT(removeRows()));
+            menu.exec(verticalHeader()->mapToGlobal(point));
+       });
+
+    connect(horizontalHeader(), &QWidget::customContextMenuRequested,
+       [this](const QPoint &point){
+            QMenu menu(this);
+            const auto selection = selectionModel()->selectedColumns();
+            const auto & [min, max] = selectedMinMaxColumns(selection);
+            if (bool isContiguous = max - min == selection.size() - 1) {
+                /*: This is shown in the context menu for the horizontal header in a spreadsheet.
+                    The number refers to how many lines are selected and will be inserted. */
+                auto insertAbove = menu.addAction(tr("Insert %n column(s) left", "", selection.size()));
+                connect(insertAbove, SIGNAL(triggered()), this, SLOT(insertColumns()));
+
+                if (max < model()->columnCount() - 1) {
+                    auto insertAfter = menu.addAction(tr("Insert %n column(s) right", "", selection.size()));
+                    connect(insertAfter, SIGNAL(triggered()), this, SLOT(insertColumnsAfter()));
+                }
+            } else {
+                auto insert = menu.addAction(tr("Insert %n non-contiguous column(s)", "", selection.size()));
```

Comment:
I don't think this can be just one column, so you don't need the parenthesis around the "(s)".